File: java/code_injection/ScriptExecution.java
import jakarta.servlet.http.*;
import javax.script.*;
import java.io.*;
import groovy.lang.GroovyShell;

public class ScriptExecution extends HttpServlet {

    // VULN 1: ScriptEngine.eval() on user-supplied discount formula for promotional pricing
    public Object evaluateDiscountFormula(HttpServletRequest req) throws ScriptException {
        String formula = req.getParameter("discount_formula");
        ScriptEngineManager manager = new ScriptEngineManager();
        ScriptEngine engine = manager.getEngineByName("JavaScript");
        engine.put("cartTotal", 150.00);
        engine.put("itemCount", 3);
        return engine.eval(formula);
    }

    // VULN 2: Runtime.exec() with string concatenation for order report generation
    public String runReportScript(HttpServletRequest req) throws IOException {
        String reportType = req.getParameter("report_type");
        String startDate  = req.getParameter("start_date");
        String endDate    = req.getParameter("end_date");
        Process proc = Runtime.getRuntime().exec(
            "python3 /opt/shopist/scripts/generate_report.py --type " + reportType
            + " --from " + startDate + " --to " + endDate
        );
        return new String(proc.getInputStream().readAllBytes());
    }

    // VULN 3: Groovy GroovyShell.evaluate() on user-supplied shipping cost expression
    public Object evaluateShippingRule(HttpServletRequest req) {
        String shippingRule = req.getParameter("shipping_rule");
        GroovyShell shell = new GroovyShell();
        shell.setVariable("weight", 2.5);
        shell.setVariable("distance", 300);
        shell.setVariable("expedited", false);
        return shell.evaluate(shippingRule);
    }
}

File: kotlin/code_injection/ScriptExecution.kt
import org.codehaus.groovy.control.CompilerConfiguration
import groovy.lang.GroovyShell
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import javax.script.ScriptEngineManager

data class ScriptRequest(val code: String, val language: String = "groovy")
data class FormulaRequest(val formula: String, val productId: String, val basePrice: Double)
data class RuleRequest(val ruleName: String, val ruleExpression: String)

@RestController
class ScriptExecution {

    // VULN 1: ScriptEngineManager eval of user-supplied code - custom pricing rule engine
    @PostMapping("/api/pricing/eval-rule")
    fun evalPricingRule(@RequestBody request: ScriptRequest): Map<String, Any> {
        val engine = ScriptEngineManager().getEngineByName("kotlin")
            ?: ScriptEngineManager().getEngineByName("groovy")
            ?: throw IllegalStateException("No script engine available")
        val result = engine.eval(request.code)
        return mapOf("result" to (result ?: "null"), "language" to request.language)
    }

    // VULN 2: GroovyShell.evaluate on user-controlled formula - discount calculation
    @PostMapping("/api/discounts/calculate")
    fun calculateDiscount(@RequestBody request: FormulaRequest): Map<String, Any> {
        val binding = groovy.lang.Binding().apply {
            setVariable("basePrice", request.basePrice)
            setVariable("productId", request.productId)
        }
        val shell = GroovyShell(binding, CompilerConfiguration())
        val result = shell.evaluate(request.formula)
        return mapOf(
            "productId" to request.productId,
            "basePrice" to request.basePrice,
            "discountedPrice" to result
        )
    }

    // VULN 3: Runtime.exec with sh -c and user-controlled rule expression - order validation rule
    @PostMapping("/api/orders/apply-rule")
    fun applyValidationRule(@RequestBody request: RuleRequest): Map<String, Any> {
        val process = Runtime.getRuntime().exec(arrayOf("sh", "-c", request.ruleExpression))
        val output = process.inputStream.bufferedReader().readText()
        val exitCode = process.waitFor()
        return mapOf(
            "ruleName" to request.ruleName,
            "output" to output,
            "exitCode" to exitCode,
            "passed" to (exitCode == 0)
        )
    }
}
